components/eventing-controller/controllers/subscription/nats/reconciler.go

```diff
@@ -144,135 +143,156 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
 
 	r.namedLogger().Debugw("received subscription reconciliation request", "namespace", req.Namespace, "name", req.Name)
 
-	actualSubscription := &eventingv1alpha1.Subscription{}
-	result := ctrl.Result{}
+	cachedSubscription := &eventingv1alpha1.Subscription{}
 
 	// Ensure the object was not deleted in the meantime
-	err := r.Client.Get(ctx, req.NamespacedName, actualSubscription)
+	err := r.Client.Get(ctx, req.NamespacedName, cachedSubscription)
 	if err != nil {
 		return ctrl.Result{}, client.IgnoreNotFound(err)
 	}
 
 	// Handle only the new subscription
-	desiredSubscription := actualSubscription.DeepCopy()
+	subscription := cachedSubscription.DeepCopy()
 
 	// Bind fields to logger
-	log := utils.LoggerWithSubscription(r.namedLogger(), desiredSubscription)
+	log := utils.LoggerWithSubscription(r.namedLogger(), subscription)
 
-	if !desiredSubscription.ObjectMeta.DeletionTimestamp.IsZero() {
+	if !subscription.ObjectMeta.DeletionTimestamp.IsZero() {
 		// The object is being deleted
-		if utils.ContainsString(desiredSubscription.ObjectMeta.Finalizers, Finalizer) {
-			if err := r.Backend.DeleteSubscription(desiredSubscription); err != nil {
+		if utils.ContainsString(subscription.ObjectMeta.Finalizers, Finalizer) {
+			if err := r.Backend.DeleteSubscription(subscription); err != nil {
 				log.Errorw("delete subscription failed", "error", err)
 				// if failed to delete the external dependency here, return with error
 				// so that it can be retried
 				return ctrl.Result{}, err
 			}
 
 			// remove our finalizer from the list and update it.
-			desiredSubscription.ObjectMeta.Finalizers = utils.RemoveString(desiredSubscription.ObjectMeta.Finalizers, Finalizer)
-			if err := r.Client.Update(ctx, desiredSubscription); err != nil {
-				events.Warn(r.recorder, desiredSubscription, events.ReasonUpdateFailed, "Update Subscription failed %s", desiredSubscription.Name)
+			subscription.ObjectMeta.Finalizers = utils.RemoveString(subscription.ObjectMeta.Finalizers, Finalizer)
+			if err := r.Client.Update(ctx, subscription); err != nil {
+				events.Warn(r.recorder, subscription, events.ReasonUpdateFailed, "Update Subscription failed %s", subscription.Name)
 				log.Errorw("remove finalizer from subscription failed", "error", err)
-				return ctrl.Result{}, err
+				if k8serrors.IsConflict(err) {
+					return ctrl.Result{Requeue: true}, nil
+				}
 			}
 			log.Debug("remove finalizer from subscription succeeded")
 			return ctrl.Result{}, nil
 		}
 	}
 
+	// The object is not being deleted, so if it does not have our finalizer,
+	// then lets add the finalizer and update the object. This is equivalent
+	// registering our finalizer.
+	if !utils.ContainsString(subscription.ObjectMeta.Finalizers, Finalizer) {
+		subscription.ObjectMeta.Finalizers = append(subscription.ObjectMeta.Finalizers, Finalizer)
+		if err := r.Update(context.Background(), subscription); err != nil {
```

Comment:
is it worth introducing yet another update just to set the finalizer? We could continue reconciling the resource and just add the finalizer in the final object.

Reply:
It will be solved in the more general context, I've added it: https://github.com/kyma-project/kyma/issues/9766#issuecomment-1013015225
The idea of this PR was to be similar with https://github.com/kyma-project/kyma/pull/13001 and not to diverge too much